### PR TITLE
Typo squatting checklist retrieval

### PR DESCRIPTION
**Pull request by zhhyu**
Finish the typo-squatting checklist retrieval component. 

Fix https://github.com/NuGet/Engineering/issues/1594, https://github.com/NuGet/Engineering/issues/1624, https://github.com/NuGet/Engineering/issues/1645

### CODE REVIEW

File: tests/NuGetGallery.Facts/Services/TyposquattingCheckServiceFacts.cs [hunk 1]
```diff
@@ -56,33 +61,16 @@ public void CheckNotTyposquattingBySameOwnersTest()
...
-            _typosquattingOwnersDoubleCheck
-                .Setup(x => x.CanUserTyposquat("microsoft_netframework_v1", "newOwner1"))
+            _typosquattingUserService
+                .Setup(x => x.CanUserTyposquat(It.IsAny<string>(), It.IsAny<string>()))
                 .Returns(true);
-            var newService = new TyposquattingCheckService(_typosquattingOwnersDoubleCheck.Object);
-            TyposquattingCheckService.PackagesCheckList = _checkList;
+            _typosquattingPackagesCheckListService
+                .Setup(x => x.GetTyposquattingChecklist(It.IsAny<int>()))
+                .Returns(_checkList);
```
Comment (skofman1) [inline tests/NuGetGallery.Facts/Services/TyposquattingCheckServiceFacts.cs]: This call is repeated a lot. You can move it to a constructor

**Comment by skofman1** [comment]
>     public class TyposquattingCheckServiceFacts

there are more tests you can add here: example: what if checklist is empty because no packages exist yet.

---
Refers to: tests/NuGetGallery.Facts/Services/TyposquattingCheckServiceFacts.cs:11 in cabb39b. [](commit_id = cabb39bb17e7ba97f01807bd100aa39803929075, deletion_comment = False)

File: src/NuGetGallery/Services/TyposquattingCheckService.cs [hunk 2]
```diff
@@ -18,18 +23,19 @@ public class TyposquattingCheckService : ITyposquattingCheckService
             new ThresholdInfo { LowerBound = 30, UpperBound = 50, Threshold = 1 },
             new ThresholdInfo { LowerBound = 50, UpperBound = 120, Threshold = 2 }
         };
-        
-        // TODO: popular packages checklist will be implemented
-        // https://github.com/NuGet/Engineering/issues/1624
-        public static List<PackageInfo> PackagesCheckList { get; set; }
+
+        private static List<string> PackagesCheckList { get; set; } = new List<string>();
```
Comment (cristinamanum) [inline src/NuGetGallery/Services/TyposquattingCheckService.cs]: Why do these two need to be static ?

Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckService.cs]: Changed to a local variable.

File: src/NuGetGallery/Services/TyposquattingCheckService.cs [hunk 3]
```diff
@@ -18,16 +22,17 @@ public class TyposquattingCheckService : ITyposquattingCheckService
...
             new ThresholdInfo { LowerBound = 50, UpperBound = 120, Threshold = 2 }
         };
-        
-        // TODO: popular packages checklist will be implemented
-        // https://github.com/NuGet/Engineering/issues/1624
-        public static List<PackageInfo> PackagesCheckList { get; set; }
 
-        private readonly ITyposquattingUserService _userTyposquattingService;
+        private static List<string> PackagesCheckList { get; set; } = new List<string>();
+        private static ConcurrentDictionary<string, string> NormalizedPackageIdDict { get; set; } = new ConcurrentDictionary<string, string>();
```
[the remark this answers is not in the document]
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckService.cs]: Done.

File: tests/NuGetGallery.Facts/Services/TyposquattingCheckServiceFacts.cs (hunk 1 above)
Comment (zhhyu) [reply tests/NuGetGallery.Facts/Services/TyposquattingCheckServiceFacts.cs]: checklist has been refactored in another way.

File: src/NuGetGallery/Migrations/201808250104094_AddIsVerifiedIndexForPackageRegistrationsTable.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+namespace NuGetGallery.Migrations
+{
+    using System;
+    using System.Data.Entity.Migrations;
+    public partial class AddIsVerifiedIndexForPackageRegistrationsTable : DbMigration
+    {
+        public override void Up()
+        {
+            // Used for getting checklist for typosquatting
+            CreateIndex(table: "PackageRegistrations", columns: new[] { "IsVerified", "DownloadCount", "Key" }, name: "IX_PackageRegistration_IsVerified_DownloadCount_Key");
```
Comment (joelverhagen) [inline src/NuGetGallery/Migrations/201808250104094_AddIsVerifiedIndexForPackageRegistrationsTable.cs]: >Key [](start = 104, length = 3)

The column of the primary key is implicitly included in all indexes (otherwise it would be unusable for indexing on the main table...). Why is it explicit here?

File: src/NuGetGallery/Services/TyposquattingCheckService.cs [hunk 5]
```diff
@@ -2,34 +2,39 @@
...
 using System.Threading.Tasks;
+using System.Linq;
 
 namespace NuGetGallery
 {
     public class TyposquattingCheckService : ITyposquattingCheckService
     {
+        // TODO: Length of checklist will be saved in the configuration file.
+        // https://github.com/NuGet/Engineering/issues/1645
+        private static int TyposquattingCheckListLength = 20000;
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/TyposquattingCheckService.cs]: >TyposquattingCheckListLength [](start = 27, length = 28)

nit: `const`

File: src/NuGetGallery/Services/TyposquattingCheckService.cs [hunk 6]
```diff
@@ -2,34 +2,39 @@
...
+
         // TODO: Threshold parameters will be saved in the configuration file.
         // https://github.com/NuGet/Engineering/issues/1645
-        private static List<ThresholdInfo> _thresholdsList = new List<ThresholdInfo>
+        private static List<ThresholdInfo> ThresholdsList = new List<ThresholdInfo>
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/TyposquattingCheckService.cs]: >ThresholdsList [](start = 43, length = 14)

`readonly`

Comment (joelverhagen) [inline src/NuGetGallery/Services/TyposquattingCheckService.cs]: >List [](start = 23, length = 4)

`IReadOnlyList`

File: src/NuGetGallery/Services/TyposquattingCheckService.cs [hunk 7]
```diff
@@ -85,12 +94,6 @@ private static int GetThreshold(string packageId)
         }
     }
 
-    public class PackageInfo
-    {
-        public string Id { get; set; }
-        public HashSet<string> Owners { get; set; }
-    }
-
     public class ThresholdInfo
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/TyposquattingCheckService.cs]: >ThresholdInfo [](start = 17, length = 13)

nit: make this a read only type by introducing a ctor and changing properties to read only (`{ get; }`)

File: src/NuGetGallery/Services/TyposquattingCheckService.cs [hunk 8]
```diff
@@ -42,40 +47,44 @@ public bool IsUploadedPackageIdTyposquatting(string uploadedPackageId, User uplo
                 throw new ArgumentNullException(nameof(uploadedPackageOwner));
             }
 
+            var packagesCheckList = _packageRegistrationRepository.GetAll()
+                .OrderByDescending(pr => pr.IsVerified)
+                .OrderByDescending(pr => pr.DownloadCount)
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/TyposquattingCheckService.cs]: >OrderByDescending [](start = 17, length = 17)

You need `ThenBy...`, otherwise this is doing two sorts.

File: src/NuGetGallery/Services/TyposquattingCheckService.cs (hunk 5 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckService.cs]: Done

File: src/NuGetGallery/Services/TyposquattingCheckService.cs (hunk 6 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckService.cs]: Done

Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckService.cs]: Done

File: src/NuGetGallery/Services/TyposquattingCheckService.cs (hunk 7 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckService.cs]: Done

File: src/NuGetGallery/Migrations/201808250104094_AddIsVerifiedIndexForPackageRegistrationsTable.cs (hunk 4 above)
Comment (zhhyu) [reply src/NuGetGallery/Migrations/201808250104094_AddIsVerifiedIndexForPackageRegistrationsTable.cs]: The index has been updated and primary key is deleted.

File: src/NuGetGallery/Services/TyposquattingCheckService.cs [hunk 9]
```diff
@@ -2,34 +2,37 @@
...
+
         // TODO: Threshold parameters will be saved in the configuration file.
         // https://github.com/NuGet/Engineering/issues/1645
-        private static List<ThresholdInfo> _thresholdsList = new List<ThresholdInfo>
+        private static readonly IReadOnlyList<ThresholdInfo> ThresholdsList = new List<ThresholdInfo>
         {
-            new ThresholdInfo { LowerBound = 0, UpperBound = 30, Threshold = 0 },
-            new ThresholdInfo { LowerBound = 30, UpperBound = 50, Threshold = 1 },
-            new ThresholdInfo { LowerBound = 50, UpperBound = 120, Threshold = 2 }
+            new ThresholdInfo (0, 30, 0),
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/TyposquattingCheckService.cs]: nit: name parameters

File: src/NuGetGallery/Migrations/201808312215409_IX_AddIsVerifiedDownloadCountIndexForPackageRegistrationsTable.cs [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+{
+    using System;
+    using System.Data.Entity.Migrations;
+    
+    public partial class IX_AddIsVerifiedDownloadCountIndexForPackageRegistrationsTable : DbMigration
+    {
+        public override void Up()
+        {
+            // Used for getting checklist for typosquatting
+            CreateIndex(table: "PackageRegistrations", columns: new[] { "IsVerified", "DownloadCount"}, name: "IX_PackageRegistration_IsVerified_DownloadCount");
```
Comment (joelverhagen) [inline src/NuGetGallery/Migrations/201808312215409_IX_AddIsVerifiedDownloadCountIndexForPackageRegistrationsTable.cs]: >table [](start = 24, length = 5)

Consider including ID in the index so that you don't even need to hit the main table.

File: src/NuGetGallery/Services/TyposquattingCheckService.cs (hunk 9 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckService.cs]: Done

File: src/NuGetGallery/Migrations/201808312215409_IX_AddIsVerifiedDownloadCountIndexForPackageRegistrationsTable.cs (hunk 10 above)
Comment (zhhyu) [reply src/NuGetGallery/Migrations/201808312215409_IX_AddIsVerifiedDownloadCountIndexForPackageRegistrationsTable.cs]: Refactor the immigration file and create sql to include the ID.

File: src/NuGetGallery/Services/TyposquattingCheckService.cs [hunk 11]
```diff
@@ -85,16 +92,16 @@ private static int GetThreshold(string packageId)
...
-
     public class ThresholdInfo
     {
-        public int LowerBound { get; set; }
-        public int UpperBound { get; set; }
-        public int Threshold { get; set; }
+        public int LowerBound { get; }
+        public int UpperBound { get; }
+        public int Threshold { get; }
+        public ThresholdInfo(int lowerBound, int upperBound, int threshod)
```
Comment (cristinamanum) [inline src/NuGetGallery/Services/TyposquattingCheckService.cs]: nit: threshod ->threshold

Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckService.cs]: Oops, sorry for the careless!

Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckService.cs]: Done

File: src/NuGetGallery/Services/TyposquattingCheckService.cs [hunk 12]
```diff
@@ -42,38 +45,42 @@ public bool IsUploadedPackageIdTyposquatting(string uploadedPackageId, User uplo
...
                 {
-                    return;
+                    collisionPackageIds.Add(packageId);
                 }
+            });
 
-                if (TyposquattingDistanceCalculation.IsDistanceLessThanThreshold(uploadedPackageId, package.Id, threshold))
+            foreach (var packageId in collisionPackageIds)
+            {
+                if (!_userTyposquattingService.CanUserTyposquat(packageId, uploadedPackageOwner.Username))
```
Comment (cristinamanum) [inline src/NuGetGallery/Services/TyposquattingCheckService.cs]: Suggestion: In case that this call does a query to db for each packageId, if possible consider adding a method in the userTyposquattingService that does a single query to the db and gets all the ids that can be typosqatt by the user. 

I would not block the PR on this but add a task if needed to track an issue.

Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckService.cs]: Ok, Great idea! I will create another issue for this because this may lead to a complex query.
